src/app/services/baw-api/model.service.ts

```diff
@@ -0,0 +1,143 @@
+
+@Injectable()
+export class ModelService<T> extends BawApiService {
+  private subjectNext(subject: Subject<T>, data: any) {
+    subject.next(this.classBuilder(data));
+    subject.complete();
+  }
+  private subjectNextList(subject: Subject<T[]>, data: any[]) {
+    subject.next(data.map(object => this.classBuilder(object)));
+    subject.complete();
+  }
+  private subjectError(subject: Subject<T | T[]>, err: APIErrorDetails) {
+    subject.error(err);
+  }
+
+  constructor(
+    http: HttpClient,
+    config: AppConfigService,
+    @Inject(MOCK_CLASS_BUILDER) private classBuilder: (object: any) => T
+  ) {
+    super(http, config);
+  }
+
+  /**
+   * Get list of models
+   * @param path URL path
+   * @param filters Api Filters
+   * @param args URL parameter values
+   */
+  protected details(
+    path: string,
+    filters: Filters,
+    ...args: Args
+  ): Subject<T[]> {
+    const subject = new Subject<T[]>();
+    const next = (objects: object[]) => this.subjectNextList(subject, objects);
+    const error = (err: APIErrorDetails) => this.subjectError(subject, err);
+    path = this.refineUrl(path, args);
+
+    this.apiDetails(next, error, path, filters);
+
+    return subject;
+  }
+
+  /**
+   * Get individual model
+   * @param path URL path
+   * @param filters Api Filters
+   * @param args URL parameter values
+   */
+  protected show(path: string, filters: Filters, ...args: Args): Subject<T> {
+    const subject = new Subject<T>();
+    const next = (object: object) => this.subjectNext(subject, object);
+    const error = (err: APIErrorDetails) => this.subjectError(subject, err);
+    path = this.refineUrl(path, args);
+
+    this.apiDetails(next, error, path, filters);
+
+    return subject;
+  }
+
+  /**
+   * Create a new model
+   * @param path URL path
+   * @param values Form details
+   * @param args URL parameter values
+   */
+  protected new(path: string, values: any, ...args: Args): Subject<T> {
+    const subject = new Subject<T>();
+    const next = (object: object) => this.subjectNext(subject, object);
+    const error = (err: APIErrorDetails) => this.subjectError(subject, err);
+    path = this.refineUrl(path, args);
+
+    this.apiCreate(next, error, path, values);
+
+    return subject;
+  }
+
+  /**
+   * Update a model
+   * @param path URL path
+   * @param values Form details
+   * @param args URL parameter values
+   */
+  protected update(path: string, values: any, ...args: Args): Subject<T> {
+    const subject = new Subject<T>();
+    const next = (object: object) => this.subjectNext(subject, object);
+    const error = (err: APIErrorDetails) => this.subjectError(subject, err);
+    path = this.refineUrl(path, args);
+
+    this.apiUpdate(next, error, path, values);
+
+    return subject;
+  }
+
+  /**
+   * Replace URL parameters with argument values
+   * @param path URL path
+   * @param args Arguments to insert
+   */
+  private refineUrl(path: string, args: Args): string {
```

Comment:
Is it possible to reuse angular's URL templating functions? We should do so

Reply:
As far as I can tell, there is no exposed method of handling URL attributes. HttpClient only exposes a method of adding HTTPParams which are the attributes placed after the `?` in a url. Eg. `https://brokenlink?http_params_go_here`.

Reply:
https://angular.io/api/router/Router#parseurl